Fix running integration tests with spaces in it's names

azat opened the pull request:
Previously they were simply ignored, and not only them, but all the
tests in that group, here is an example [1]

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/51448/4ed462ac7834a8180f92ca7d7d3c076e687bfca9/integration_tests__asan__[4_6].html

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

azat committed:
Fix running integration tests with spaces in it's names

Previously they were simply ignored, and not only them, but all the
tests in that group, here is an example [1]

  [1]: https://s3.amazonaws.com/clickhouse-test-reports/51448/4ed462ac7834a8180f92ca7d7d3c076e687bfca9/integration_tests__asan__[4_6].html

Note, that I've replaced PYTEST_OPTS with handled by pytest itself
PYTEST_ADDOPTS, since "sh -c $PYTEST_OPTS" does not allows you to
preserve the arguments correctly.

v5: use double quotes for PYTEST_ADDOPTS value
Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

robot-ch-test-poll3 commented:
<!-- automatic status comment for PR #51514 from azat/ClickHouse:tests/fix-integration-tests -->
This is an automated comment for commit d4135cdac33e3f0e037374c04d7f113972c26e75 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51514/d4135cdac33e3f0e037374c04d7f113972c26e75/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

qoega commented on tests/integration/runner:
```diff
@@ -415,8 +406,9 @@ if __name__ == "__main__":
-        opts=" ".join(args.pytest_args).replace("'", "\\'"),
-        tests_list=" ".join(args.tests_list),
+        opts=" ".join(
+            map(lambda x: shlex.quote(x).replace("'", "\\'"), args.pytest_args)
```
I have a feeling that there should be easier way than `.quote(x).replace("'", "\\'")`, but who cares.

azat replied:
There are more problems with this, but anyway after other fixes it can be converted to a plain `shlex.quote`.

azat committed:
Add _gen to ignore (generated by pytest)

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: use -- over quotes on and on in ci-runner.py

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
tests: do not use --format if --quiet is set in integration runner

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
>[Integration tests (asan) [6/6] — fail: 20, passed: 285, flaky: 0](https://s3.amazonaws.com/clickhouse-test-reports/51514/2dad15c721c6b07a2e8cf1b2550c37cf0a0dd3e5/integration_tests__asan__[6_6].html)

One of jobs got killed:

https://s3.amazonaws.com/clickhouse-test-reports/51514/2dad15c721c6b07a2e8cf1b2550c37cf0a0dd3e5/integration_tests__asan__[6_6]/integration_run_parallel2_0.log

```
[gw1] [ 81%] PASSED test_merge_tree_s3/test.py::test_table_manipulations[node] Killed
```

alexey-milovidov commented:
What has happened with the test for `test_postgresql_replica_database_engine_1`?

azat commented:
All tests had been failed because of the runner got killed (see above):

```
[gw1] [ 81%] PASSED test_merge_tree_s3/test.py::test_table_manipulations[node] Killed
```

alexey-milovidov commented:
This PR was expected to find a failed test for UDF, but it didn't: https://github.com/ClickHouse/ClickHouse/pull/51964

alexey-milovidov commented:
Interesting, why the runner was killed?

alexey-milovidov commented:
What's wrong with

> Upgrade check (tsan) — Possible deadlock on shutdown (see gdb.log)

?

alexey-milovidov commented:
@azat I cannot find `executable_user_defined_function/` in the list of passed tests.

alexey-milovidov commented:
@azat This group https://s3.amazonaws.com/clickhouse-test-reports/51514/7e8ccade3517f55e111c6dbfac653dcd7ed18019/integration_tests__release__[1_4].html
has parallel runs №1, 2, 3, and 5, but not 4.

alexey-milovidov reviewed (changes requested):
.

azat commented:
>@azat I cannot find executable_user_defined_function/ in the list of passed tests.

Nice catch, there was an error due to single quotes in quotes:

https://s3.amazonaws.com/clickhouse-test-reports/51514/7e8ccade3517f55e111c6dbfac653dcd7ed18019/integration_tests__release__[4_4]/integration_run_parallel1_0.log
```
docker: invalid reference format.
```

>@azat This group [https://s3.amazonaws.com/clickhouse-test-reports/51514/7e8ccade3517f55e111c6dbfac653dcd7ed18019/integration_tests__release__[1_4].html](https://s3.amazonaws.com/clickhouse-test-reports/51514/7e8ccade3517f55e111c6dbfac653dcd7ed18019/integration_tests__release__%5B1_4%5D.html)
has parallel runs №1, 2, 3, and 5, but not 4.

Same:

https://s3.amazonaws.com/clickhouse-test-reports/51514/7e8ccade3517f55e111c6dbfac653dcd7ed18019/integration_tests__release__[1_4]/integration_run_parallel4_0.log
```
/bin/sh: 1: Syntax error: "(" unexpected
```

Both had be addressed

alexey-milovidov commented:
Now it solves the problem!
But we have to fix the newly found forgotten tests...

azat committed:
Allow test names with spaces in it

It had been fixed and now it is allowed.

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat commented:
> test_extreme_deduplication/test.py::test_deduplication_window_in_seconds

- https://github.com/ClickHouse/ClickHouse/pull/52085

azat commented:
>Integration tests (asan) [4/6] — fail: 1, passed: 374, flaky: 0
>Integration tests (release) [4/4] — fail: 1, passed: 525, flaky: 0
>Integration tests (tsan) [4/6] — fail: 1, passed: 374, flaky: 0


- `test_extreme_deduplication` - #52085 

>Integration tests (asan, analyzer) [1/6] — fail: 34, passed: 261, flaky: 0

- Differences in exception names - I.e. `Unknown function MySum` vs `Function with name 'MySum' does not exists`
- Somewhere the server crashes (since optimize_skip_unused_shards leads to SIGSEGV now, I was going to look at this after https://github.com/ClickHouse/ClickHouse/pull/51037)
- `TCPHandler: Code: 10. DB::Exception: Not found column tcpPort() in block.`
- `2023.07.13 13:39:08.509341 [ 8 ] {} <Error> Application: DB::Exception: Setting allow_experimental_analyzer is neither a builtin setting nor started with the prefix 'custom_' registered for user-defined settings`

>Integration tests (release) [2/4] — fail: 8, passed: 565, flaky: 0

- `test_lost_part`

```
E           AssertionError: Still have something in replication queue:
E             Row 1:
E             ──────
E             database:               default
E             table:                  mt3
E             replica_name:           node2
E             position:               0
E             node_name:              queue-0000000000
E             type:                   GET_PART
E             create_time:            2023-07-13 14:07:07
E             required_quorum:        0
E             source_replica:         node1
E             new_part_name:          8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0
E             parts_to_merge:         []
E             is_detach:              0
E             is_currently_executing: 0
E             num_tries:              13108
E             last_exception:         
E             last_exception_time:    1970-01-01 00:00:00
E             last_attempt_time:      2023-07-13 14:13:07
E             num_postponed:          13593
E             postpone_reason:        Not executing log entry queue-0000000000 for part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0 because it is not disjoint with part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1 that is currently executing.
E             last_postpone_time:     2023-07-13 14:13:07
E             merge_type:             
E             
E             Row 2:
E             ──────
E             database:               default
E             table:                  mt3
E             replica_name:           node2
E             position:               1
E             node_name:              queue-0000000001
E             type:                   MUTATE_PART
E             create_time:            2023-07-13 14:07:07
E             required_quorum:        0
E             source_replica:         node1
E             new_part_name:          8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1
E             parts_to_merge:         ['8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0']
E             is_detach:              0
E             is_currently_executing: 1
E             num_tries:              13108
E             last_exception:         
E             last_exception_time:    1970-01-01 00:00:00
E             last_attempt_time:      2023-07-13 14:13:07
E             num_postponed:          13107
E             postpone_reason:        Not executing log entry queue-0000000001 for part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0_1 because it is not disjoint with part 8b8f0fede53df97513a9fb4cb19dc1e4_0_0_0 that is currently executing.
E             last_postpone_time:     2023-07-13 14:13:07
E             merge_type:             
E             
E           assert False
```

>Integration tests (tsan) [5/6] — fail: 1, passed: 521, flaky: 2

- `test_storage_s3/test.py::test_url_reconnect_in_the_middle - Exception: ...` - https://github.com/ClickHouse/ClickHouse/issues/51345

azat committed:
Update analyzer integration broken tests list

After proper parsing more failed tests pops up:
- https://s3.amazonaws.com/clickhouse-test-reports/51514/aa90e11241a97e0fd3f36c5ea0b57db50e299d2d/integration_tests__asan__analyzer__[4_6].html
- https://s3.amazonaws.com/clickhouse-test-reports/51514/aa90e11241a97e0fd3f36c5ea0b57db50e299d2d/integration_tests__asan__analyzer__[5_6].html

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Make psql tests in test_odbc_interaction more robust to other test failures

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

azat committed:
Fix reporting some BROKEN tests as ERROR

Signed-off-by: Azat Khuzhin <a.khuzhin@semrush.com>

antonio2368 commented on tests/integration/runner:
```diff
@@ -407,8 +406,14 @@ if __name__ == "__main__":
-    pytest_opts = " ".join(args.pytest_args).replace("'", "\\'")
-    tests_list = " ".join(args.tests_list)
+    # NOTE: since pytest options is in the argument value already we need to additionally escape '"'
+    pytest_opts = " ".join(
```
@azat this broke sending extra arguments to pytest like `-k` because everything is put under the same quote and parsed as test name.

azat replied:
It should work, maybe you are using old env variable name? The new one is `PYTEST_ADDOPTS`

antonio2368 replied:
did you try running an integration test using runner and filtering, e.g. `test_keeper_back_to_back -k test_concurrent_watches`?

AFAIS, the entire `pytest_args` are quoted.
